exporter/internal/queue/default_batcher.go

```diff
@@ -55,76 +57,102 @@ func (qb *defaultBatcher) resetTimer() {
+	var mergeSplitErr error
+	if qb.currentBatch == nil {
+		reqList, mergeSplitErr = req.MergeSplit(ctx, qb.batchCfg.MaxSizeConfig, nil)
+		if mergeSplitErr != nil || len(reqList) == 0 {
+			done.OnDone(mergeSplitErr)
 			qb.currentBatchMu.Unlock()
 			return
 		}
 
-		// If there was a split, we flush everything immediately.
-		if reqList[0].ItemsCount() >= qb.batchCfg.MinSizeItems || len(reqList) > 1 {
-			qb.currentBatch = nil
-			qb.currentBatchMu.Unlock()
-			for i := 0; i < len(reqList); i++ {
-				qb.flush(ctx, reqList[i], []exporterqueue.DoneCallback{done})
-				// TODO: handle partial failure
-			}
-			qb.resetTimer()
-		} else {
-			qb.currentBatch = &batch{
-				req:   reqList[0],
-				ctx:   ctx,
-				dones: []exporterqueue.DoneCallback{done},
-			}
-			qb.currentBatchMu.Unlock()
+		// If more than one flush is required for this request, call done only when all flushes are done.
+		if len(reqList) > 1 {
+			done = newRefCountDone(done, int64(len(reqList)))
 		}
-	} else {
-		if qb.currentBatch == nil || qb.currentBatch.req == nil {
-			qb.resetTimer()
-			qb.currentBatch = &batch{
-				req:   req,
-				ctx:   ctx,
-				dones: []exporterqueue.DoneCallback{done},
-			}
-		} else {
-			// TODO: consolidate implementation for the cases where MaxSizeConfig is specified and the case where it is not specified
-			mergedReq, mergeErr := qb.currentBatch.req.MergeSplit(qb.currentBatch.ctx, qb.batchCfg.MaxSizeConfig, req)
-			if mergeErr != nil {
-				done(mergeErr)
-				qb.currentBatchMu.Unlock()
-				return
-			}
+
+		// We have at least one result in the reqList. Last in the list may not have enough data to be flushed.
+		// Find if it has at least MinSizeItems, and if it does then move that as the current batch.
+		lastReq := reqList[len(reqList)-1]
+		if lastReq.ItemsCount() < qb.batchCfg.MinSizeItems {
+			// Do not flush the last item and add it to the current batch.
+			reqList = reqList[:len(reqList)-1]
 			qb.currentBatch = &batch{
-				req:   mergedReq[0],
-				ctx:   qb.currentBatch.ctx,
-				dones: append(qb.currentBatch.dones, done),
+				ctx:  ctx,
+				req:  lastReq,
+				done: multiDone{done},
 			}
+			qb.resetTimer()
 		}
 
-		if qb.currentBatch.req.ItemsCount() >= qb.batchCfg.MinSizeItems {
-			batchToFlush := *qb.currentBatch
-			qb.currentBatch = nil
-			qb.currentBatchMu.Unlock()
+		qb.currentBatchMu.Unlock()
+		for i := 0; i < len(reqList); i++ {
+			qb.flush(ctx, reqList[i], done)
+		}
 
-			// flush() blocks until successfully started a goroutine for flushing.
-			qb.flush(batchToFlush.ctx, batchToFlush.req, batchToFlush.dones)
+		return
+	}
+
+	reqList, mergeSplitErr = qb.currentBatch.req.MergeSplit(ctx, qb.batchCfg.MaxSizeConfig, req)
+	// If failed to merge signal all Done callbacks from current batch as well as the current request and reset the current batch.
+	if mergeSplitErr != nil || len(reqList) == 0 {
+		// TODO: investigate why is ok to keep old data here, this is the current behavior in bunch of tests.
```

Comment:
I believe this error should be associated with the new incoming request only  We can just send the batch accumulated in `qb. current batch` instead. It must be valid by itself. I'm not sure why we would need to drop it.

Reply:
> I believe mergeSplitErr should be associated with the new incoming request only. 

I could not find anywhere in the documentation that MergeSplit should not "touch" the first request passed if fails. Since we do pass `currentBatch.req` to that function, unless we have a contract that it is "valid" not sure we can continue to use it.

I left this unchanged, but we need to document the expected behavior, and if we say that should not affect passed request then we can remove this.

Reply:
Filed https://github.com/open-telemetry/opentelemetry-collector/issues/12268, and I remove this TODO.